haskell/data_structures/bloom_filter.hs

```diff
@@ -0,0 +1,48 @@
+module BloomFilter (empty, insert, member, BloomFilter.null) where
+
+import System.Random (random, StdGen)
+import qualified Data.BitSet as BitSet (BitSet, empty, insert, member, null, size)
+import Data.Hashable (hashWithSalt)
+
+-- n: expected number of items in the Bloom Filter
+-- p: acceptable probability of a false positive
+-- m: max number of bits the Bloom Filter will use
+-- k: number of hashing functions
+
+data BloomFilter = BloomFilter {
+  n :: Int, p :: Float, bitset :: BitSet.BitSet Int, m :: Int, k :: Int, hashSeed :: Int
+} deriving (Eq, Show)
+
+getMaxSize :: Int -> Float -> Int
+getMaxSize n p = abs $ ceiling $ fromIntegral n * (log p) / (log (1 / (log 2 ^ 2)))
+
+getNumberOfHashFunctions :: Int -> Int -> Int
+getNumberOfHashFunctions n m = round $ fromIntegral (m `div` n) * log 2
+
+empty :: Int -> Float -> StdGen -> BloomFilter
+empty n p randGen =
+  let m = getMaxSize n p
+      k = getNumberOfHashFunctions n m
+      hashSeed = fst $ random randGen
+  in BloomFilter n p BitSet.empty m k hashSeed
+
+null :: BloomFilter -> Bool
+null = BitSet.null . bitset
+
+getHashes :: Show a => BloomFilter -> a -> [Int]
+getHashes bloomFilter elem =
+  let str = show elem
+      seed = hashSeed bloomFilter
+  in map ((flip hashWithSalt str) . (seed +)) [1..(k bloomFilter)]
+
+insert :: Show a => a -> BloomFilter -> BloomFilter
+insert elem bloomFilter =
+  let hashes = getHashes bloomFilter elem
+      newBitSet = foldl (flip BitSet.insert) (bitset bloomFilter) hashes
+  in bloomFilter { bitset = newBitSet }
+
+member :: Show a => a -> BloomFilter -> Bool
```

Comment:
Does a return value of `true` mean that elem is definitely not in the set?
Does haskell support enumerations? Would it be "haskellic" to express the result as
`NOT_IN_SET` and `MAY_BE_IN_SET` instead?
How about comments on the output of each function for ease of learning?

Reply:
> Does a return value of true mean that elem is definitely not in the set?

A return value of `false` (not `true`) does mean that the element is definitely not in the set (a bloom filter cannot yield a false negative. However, it may yield false positives, which is why I called that same method `mayContain` in the Scala implementation of the Bloom filter.
I am not fan of outputting an enum (and I don't think that's a very common pratice), but I have nothing against having the method name reflect the possibility of a false positive, like in the Scala implementation.
However, the standard for `contains` in Haskell is `member`, so I wasn't sure how to call it... `mayBeMember`? But then it's kind of ugly.
Also, one could make the argument that a `mayBeMember` should return a `Maybe Bool` (`Option(Boolean)` in Scala).
Following that argmuent, one could argue that it's the client's responsibility to know that the `member` method of a bloom filter can yield a false positive, and the API does not necessarily have to reflect that :thinking:

Reply:
> How about comments on the output of each function for ease of learning?

Ease of learning of Haskell or Bloom Filters?
Cuz as far as Haskell is concerned, I believe I can't be much more explicit than Haskell's awesome type signatures (the line right above the actual function definition)

Reply:
> > How about comments on the output of each function for ease of learning?
> 
> Ease of learning of Haskell or Bloom Filters?
> Cuz as far as Haskell is concerned, I believe I can't be much more explicit than Haskell's awesome type signatures (the line right above the actual function definition)

Bloom Filters. Like a comment that `member` returning true could be a false positive. Or how about calling it `candidateMember`(if it is not a candidate, it's definitely not in the set. If it is a candidate, it _may_ be in the set.)

Reply:
`mayBeMember` might still be confusing because it sounds like any outcome is probabilistic.
(If it returns false, it can be read as _may not be_ in the set)

Reply:
It seems that in other implementations, the API doesn't reflect the "proability" aspect of the member query: https://github.com/bos/bloomfilter/blob/master/Data/BloomFilter.hs#L192
I guess it's understood that if you're using a bloom filter, you know how it works... :man_shrugging: 
I'm quite torn tbh. I guess I can add a lengthy comment above my `member` function, just like the guy did in the file I linked

Reply:
I think if someone is reading this repo, they might be learning a new language and/or learning new basic algorithms. A comment on what `member` means would help.